common/src/View/ScaleObjectsToolPage.cpp

```diff
@@ -42,39 +42,89 @@ namespace TrenchBroom {
         m_document(document) {
             createGui();
         }
-        
+
+        void ScaleObjectsToolPage::activate() {
+            const auto document = lock(m_document);
+            const auto suggestedSize = document->hasSelectedNodes() ? document->selectionBounds().size() : vm::vec3::zero;
+
+            m_sizeTextBox->SetValue(StringUtils::toString(suggestedSize));
+            m_factorsTextBox->SetValue("1.0 1.0 1.0");
+        }
+
         void ScaleObjectsToolPage::createGui() {
-            wxStaticText* text = new wxStaticText(this, wxID_ANY, "Scale objects by");
-            m_scaleFactors = new wxTextCtrl(this, wxID_ANY, "1.0 1.0 1.0");
+            MapDocumentSPtr document = lock(m_document);
+
+            wxStaticText* text = new wxStaticText(this, wxID_ANY, "Scale objects");
+
+            m_book = new wxSimplebook(this);
+            m_sizeTextBox = new wxTextCtrl(m_book, wxID_ANY, "", wxDefaultPosition, wxDefaultSize, wxTE_PROCESS_ENTER);
+            m_factorsTextBox = new wxTextCtrl(m_book, wxID_ANY, "", wxDefaultPosition, wxDefaultSize, wxTE_PROCESS_ENTER);
+            m_book->AddPage(m_sizeTextBox, "");
+            m_book->AddPage(m_factorsTextBox, "");
+
+            m_sizeTextBox->Bind(wxEVT_TEXT_ENTER, &ScaleObjectsToolPage::OnApply, this);
+            m_factorsTextBox->Bind(wxEVT_TEXT_ENTER, &ScaleObjectsToolPage::OnApply, this);
+
+            const wxString choices[] = { "to size", "by factors" };
+            m_scaleFactorsOrSize = new wxChoice(this, wxID_ANY, wxDefaultPosition, wxDefaultSize, 2, choices);
+            m_scaleFactorsOrSize->Bind(wxEVT_CHOICE, [&](wxCommandEvent& event){
+                    const auto selection = m_scaleFactorsOrSize->GetSelection();
+                    if (selection != wxNOT_FOUND) {
+                        m_book->SetSelection(static_cast<size_t>(selection));
```

Comment:
Just out of interest, why not use the same text box for both absolute and relative scaling?

Reply:
I wanted to copy the behaviour of the rotate tool where it only fills in the suggested scale size when you activate the tool (the rotate tool only fills in the rotation origin when you activate the tool).

It gets a bit messy with only one text box if you allow the tool to activate with the "scale by factors" mode selected, because you'd have to cache the selection bounds size, and then fill the text box with that value.

That's another question: currently I don't reset the scaling mode when the tool is activated, should we do that, or just leave it on whatever mode the user last had it set to?

Reply:
I see. Ok! Regarding the scale mode, I‘d just leave it. You can be sure some people will want this to persist as a preference even ;-)